File: Awesome.AI.Source/Awesome.AI/Core/Spaces/Unit.cs
using Awesome.AI.Common;
using Awesome.AI.Core.Internals;
using Awesome.AI.CoreSystems;
using Awesome.AI.Interfaces;
using Awesome.AI.Source.Awesome.AI.Common;
using Awesome.AI.Variables;
using static Awesome.AI.Variables.Enums;

namespace Awesome.AI.Core.Spaces
{
    public class UNIT
    {
        /*
         * maybe later there will be other types, like: SOUND, IMAGE, VIDEO
         * */

        public Ticket ticket { get; set; }
        private UNITTYPE unit_type { get; set; }
        public LONGTYPE ld_type { get; set; }
        public DateTime created { get; set; }
        public string guid { get; set; }
        public double credits { get; set; }
        public double reward { get; set; }        
        public double trace { get; set; }
                
        private TheMind mind;
        private UNIT() { }
        public UNIT(TheMind mind)
        {
            this.mind = mind;
        }

        private GPTVector2D vect = new GPTVector2D();
                
        public double UIget(string ax)
        {
            if (IsIDLE())
                return 50.0d;

            switch (ax)
            {
                case "will": return vect.xx;
                case "attention": return vect.yy;
                default: throw new Exception("Unit, UIget");
            }
        }

        public void UIset(string ax, double val)
        {
            switch (ax)
            {
                case "will": vect.xx = val; break;
                case "attention": vect.yy = val; break;
                default: throw new Exception("Unit, UIset");
            }            
        }

        private double h_index { get; set; }
        public double HI
        {
            get => IsIDLE() ? 50.0 : h_index;
            set { h_index = value; }
        }

        private string data { get; set; }//data
        public string Data
        {
            get
            {
                if (data != "DATA" && data != CONST.QUICK)
                    return data;

                switch (data)
                {
                    case "DATA":
                        string sub = mind.hub.GetSubject(this);

                        if (sub == "init")
                            return "";

                        double _i = mind.mech.ms.dv_sym_100.Norm0DV(mind);
                        string idx = $"{_i.Index(mind)}";

                        Lookup lookup = new Lookup();

                        string _data = lookup.GetDATA(mind, idx, sub);

                        return _data;
                    case "QUICK":
                        double index = UIget("will");

                        return index < 50.0d ? CONST.Q_YES : CONST.Q_NO;
                    default:
                        throw new Exception("Unit, Data");
                }
            }
            set { data = value; }
        }

        public string Root
        {
            get
            {
                List<UNIT> list = mind.hub.UnitsPerOccupasionc().OrderBy(x => x.created).ToList();

                if (!list.Any())
                    return "";

                int idx = list.IndexOf(this) + 1;
                string res = "_" + mind.hub.GetSubject(this) + idx;

                return res;
            }
        }

        public double Variable
        {
            get
            {
                IMechanics mech = mind.mech;
                double res = 0.0d;
                switch (mech.type)
                {
                    case MECHANICS.CIRCUIT_1_LOW:
                        throw new Exception("UNIT, Variable");
                    case MECHANICS.CIRCUIT_2_LOW:
                        res = UIget("will").LowZero(); break;
                    case MECHANICS.TUGOFWAR_LOW:
                        res = UIget("will").HighZero(); break;
                    case MECHANICS.BALLONHILL_LOW:
                        res = UIget("will").LowZero(); break;
                    default: throw new Exception("UNIT, Variable");
                }

                switch (CONST.transfer)
                {
                    case TRANSFER.NONE: break;
                    case TRANSFER.LOGISTIC: res = mind.calc.Logistic(res * 0.1d - 5.0d) * 100.0d; break;
                    case TRANSFER.OTHER: throw new NotImplementedException("Unit, Variable 1");
                    default : throw new Exception("Unit, Variable 2");
                }

                return res;
            }
        }

        public bool IsValid
        {
            get
            {
                switch (mind.bot.validation)
                {
                    case VALIDATION.BOTH:
                        return mind._internal.Valid(this) && mind._external.Valid(this);
                    case VALIDATION.EXTERNAL:
                        return mind._external.Valid(this);
                    case VALIDATION.INTERNAL:
                        return mind._internal.Valid(this);
                    default:
                        throw new Exception("IsValid");
                }
            }
        }

        public static UNIT Create(TheMind mind, string h_guid, double[] index, string data, string ticket, UNITTYPE ut, LONGTYPE lt)
        {
            //make sure some time has gone before creating a new unit
            if (ut != UNITTYPE.QDECISION)
                "hello world!".BusyWait(1, mind);

            DateTime create = DateTime.Now;
            Random rand = new Random();
            Lookup lookup = new Lookup();
            
            UNIT _w = new UNIT() { mind = mind, created = create, guid = h_guid, data = data, unit_type = ut, ld_type = lt };

            for (int i = 0; i < CONST.AXIS_MAX; i++)
                _w.UIset(CONST.AXES[i], index[i]);

            ticket = ticket != "" ? ticket : "NOTICKET";
            _w.ticket = new Ticket(ticket);

            double _r1 = rand.NextDouble();

            _w.trace = 1.0d;
            _w.credits = CONST.MAX_CREDIT;
            _w.h_index = _r1 * CONST.MAX_HUBSPACE;

            return _w;
        }

        public static UNIT CreateIdle(TheMind mind)
        {
            return Create(mind, "GUID", [-1d, -1d], "IDLE", "NONE", UNITTYPE.IDLE, LONGTYPE.NONE);
        }

        public static UNIT CreateQuick(TheMind mind, string name, double[] dex)
        {
            return Create(mind, "GUID", dex, name, "NONE", UNITTYPE.QDECISION, LONGTYPE.NONE);
        }

        public GPTVector2D ToVector() 
        {
            return new GPTVector2D(UIget(CONST.AXES[0]), UIget(CONST.AXES[1]), null, null);
        }

        public void Update(GPTVector2D v_near)
        {
            UpdateTRA();
            UpdateREW();
            UpdateHUB();
            UpdateUNT(v_near);
        }

        private void UpdateTRA()
        {
            if (IsDECISION())
                return;

            List<UNIT> units = mind.hub.UnitsPerOccupasionc();

            foreach (UNIT unit in units)
            {
                if (unit.guid == this.guid)
                    unit.trace = CONST.DECAY * unit.trace + 1.0d;
                else
                    unit.trace = CONST.DECAY * unit.trace + 0.0d;                    
            }
        }

        private void UpdateREW()
        {
            if (!mind.reward)
                return;

            if (IsDECISION())
                return;

            reward += 1.0 * trace;

            List<UNIT> units = mind.hub.UnitsPerOccupasionc();

            double max = units.Max(x => x.reward);
            
            foreach (UNIT unit in units)
            {
                if (unit.guid == this.guid)
                    continue;

                double reward_norm = mind.calc.Normalize(reward, 0.0d, max, 0.0d, 1.0d);

                if (reward_norm < CONST.EPSILON1)
                    mind.access.UNITS_REM(unit);                
            }
        }

        private void UpdateHUB()
        {
            if (IsDECISION())
                return;

            mind.reward = false;

            List<UNIT> units = mind.hub.UnitsPerOccupasionc();

            string sub = mind.hub.GetSubject(this);

            double index = mind.hub.GetIndex(sub);

            if (index == -1)
                return;

            double max = units.Max(x => x.reward);

            double reward_norm = mind.calc.Normalize(reward, 0.0d, max, 0.0d, 1.0d);

            double effective_gamma = CONST.GAMMA * (1.0d + reward_norm);

            HI += HI < index ? effective_gamma : -effective_gamma;

            mind.hub.AdjustWeights(sub, effective_gamma * 0.1d);
        }

        private void UpdateUNT(GPTVector2D v_near)
        {
            if (mind.STATE == STATE.QUICKDECISION)
                return;

            if (Add2D(v_near))
                return;

            Adjust();
        }

        private bool Add2D(GPTVector2D v_near)
        {
            int axis_count = CONST.AXES.Length;
            Lookup lookup = new Lookup();
            int count_units = mind.access.UNITS_ALL().Count;
            int count_hubs = lookup.CountHUBS(mind.mindtype);

            //double avg_area = (100.0 * 100.0) / count;
            //double avg_radius = Math.Sqrt(avg_area / Math.PI);

            //if (dist < avg_radius)
            //    return false;

            if (count_units > (CONST.MAX_UNITS * count_hubs))
                return false;

            double[] xx = [-1d, -1d];
            double[] yy = [-1d, -1d];
            
            xx[0] = Math.Clamp(v_near.xx - CONST.ALPHA, CONST.MIN, CONST.MAX);
            xx[1] = Math.Clamp(v_near.xx + CONST.ALPHA, CONST.MIN, CONST.MAX);
            
            yy[0] = Math.Clamp(v_near.yy - CONST.ALPHA, CONST.MIN, CONST.MAX);
            yy[1] = Math.Clamp(v_near.yy + CONST.ALPHA, CONST.MIN, CONST.MAX);
            
            double[][] axis = new double[axis_count][];
            
            axis[0] = xx;
            axis[1] = yy;
            
            mind.access.UNITS_ADD(this, axis, axis_count);

            return true;
        }
                
        private void Adjust()
        {
            double rnd = mind.rand.MyRandomDouble(10)[5];

            GPTVector2D func = new GPTVector2D();
            GPTVector2D vec = ToVector();
            GPTVector2D dir = mind.down.FlipUnit(vec);
            dir = func.Mul(dir, rnd * CONST.ETA);
            GPTVector2D _new = func.Add(vec, dir);

            if (_new.xx <= CONST.MIN) _new.xx = CONST.MIN;
            if (_new.yy >= CONST.MAX) _new.yy = CONST.MAX;

            switch (_new.xx)
            {
                case < CONST.MIN + CONST.MIN: mind.access.UNITS_REM(this); break;
                case > CONST.MAX - CONST.MIN: mind.access.UNITS_REM(this); break;
                default: UIset("will", _new.xx); return;
            }

            switch (_new.yy)
            {
                case < CONST.MIN + CONST.MIN: mind.access.UNITS_REM(this); break;
                case > CONST.MAX - CONST.MIN: mind.access.UNITS_REM(this); break;
                default: UIset("attention", _new.yy); return;
            }
        }

        //private void Remove(double near)
        //{
        //    double low = near - CONST.ALPHA;
        //    double high = near + CONST.ALPHA;

        //    mind.access.UNITS_REM(this, low, high);
        //}

        public bool IsUNIT() => unit_type == UNITTYPE.JUSTAUNIT;

        public bool IsIDLE() => unit_type == UNITTYPE.IDLE;

        public bool IsDECISION() => unit_type == UNITTYPE.LDECISION;

        public bool IsQDECISION() => unit_type == UNITTYPE.QDECISION;
    }
}

File: Awesome.AI.Python/Core/Spaces/Unit.py
import random
from dataclasses import dataclass, field
from datetime import datetime, timezone
from uuid import uuid4
from awesome_ai.Common import GPTVector2D, GPTVector6D
from awesome_ai.Common.MyExtensions import HighZero, Index, LowZero
from awesome_ai.Core.Internals.Lookup import Lookup
from awesome_ai.CoreSystems.Environment import Ticket
from awesome_ai.Variables.Constants import CONST
from awesome_ai.Variables.Enums import LONGTYPE, MECHANICS, STATE, TRANSFER, UNITTYPE, VALIDATION


@dataclass
class UNIT:
    ticket: object = None
    unit_type: UNITTYPE = UNITTYPE.JUSTAUNIT
    ld_type: LONGTYPE = LONGTYPE.NONE
    created: datetime = field(default_factory=lambda: datetime.now(timezone.utc))
    guid: str = field(default_factory=lambda: str(uuid4()))
    credits: float = 0.0
    reward: float = 0.0
    base_props: dict = field(default_factory=dict)
    h_index: float = 0.0
    data: str = ""
    mind: object = field(default=None, repr=False, compare=False)

    def UIget(self, ax): return 50.0 if self.IsIDLE() else self.base_props[ax]
    def UIset(self, ax, val): self.base_props[ax] = float(val)
    def HIget(self): return 50.0 if self.IsIDLE() else self.h_index
    def HIset(self, val): self.h_index = float(val)

    @staticmethod
    def Create(mind, h_guid, index, data, ticket, ut, lt):
        unit = UNIT(
            ticket=Ticket(ticket or "NOTICKET"), unit_type=ut, ld_type=lt,
            guid=h_guid, data=data, mind=mind, reward=1.0,
            h_index=random.random() * CONST.MAX_HUBSPACE,
        )
        if CONST.USE_CREDITS:
            unit.credits = CONST.MAX_CREDIT
        for axis, value in zip(CONST.AXES, index):
            unit.UIset(axis, value)
        return unit

    @staticmethod
    def CreateIdle(mind): return UNIT.Create(mind, "GUID", [-1.0, -1.0], "IDLE", "NONE", UNITTYPE.IDLE, LONGTYPE.NONE)
    @staticmethod
    def CreateQuick(mind, name, dex): return UNIT.Create(mind, "GUID", dex, name, "NONE", UNITTYPE.QDECISION, LONGTYPE.NONE)
    def ToVector2D(self): return GPTVector2D(self.UIget(CONST.AXES[0]), self.UIget(CONST.AXES[1]))
    def ToVector6D(self): return GPTVector6D(*(self.UIget(k) for k in CONST.AXES))
    def Update2D(self, vector):
        self.UpdateTRA(); self.UpdateREW(); self.UpdateHUB()
        if self.mind.STATE == STATE.QUICKDECISION: return
        if self.Add2D(vector): return
        self.Adjust2D()
    def Update6D(self, vector):
        self.UpdateTRA(); self.UpdateREW(); self.UpdateHUB()
        if self.mind.STATE == STATE.QUICKDECISION: return
        if self.Add6D(vector): return
        self.Adjust6D()
    def UpdateTRA(self):
        if self.IsDECISION(): return
        for unit in self.mind.hub.UnitsPerOccupasionc(): unit.reward = CONST.DECAY * unit.reward + (1.0 if unit.guid == self.guid else 0.0)
    def UpdateREW(self):
        if not self.mind.reward or self.IsDECISION(): return
        units = self.mind.hub.UnitsPerOccupasionc(); maximum = max((unit.reward for unit in units), default=0)
        if maximum: [self.mind.access.UNITS_REM(unit) for unit in list(units) if unit.guid != self.guid and self.mind.calc.Normalize(self.reward, 0, maximum, 0, 1) < CONST.EPSILON1]
    def UpdateHUB(self):
        if self.IsDECISION(): return
        self.mind.reward = False; sub = self.mind.hub.GetSubject(self); index = self.mind.hub.GetIndex(sub)
        if index < 0: return
        units = self.mind.hub.UnitsPerOccupasionc(); maximum = max((unit.reward for unit in units), default=1)
        gamma = CONST.GAMMA * (1 + self.mind.calc.Normalize(self.reward, 0, maximum, 0, 1))
        self.HIset(self.HIget() + (gamma if self.HIget() < index else -gamma)); self.mind.hub.AdjustWeights(sub, gamma * .1)
    def _can_add(self): return len(self.mind.access.UNITS_ALL()) <= CONST.MAX_UNITS * Lookup().CountHUBS(self.mind.mindtype)
    def _axis(self, values): return [[max(CONST.MIN, value - CONST.ALPHA), min(CONST.MAX, value + CONST.ALPHA)] for value in values]
    def Add2D(self, vector):
        if not self._can_add(): return False
        # C# allocates one range per configured axis, while its two first
        # dimensions are populated from the 2D vector.
        axis = self._axis((vector.xx, vector.yy)) + [[CONST.MIN, CONST.MAX] for _ in CONST.AXES[2:]]
        self.mind.access.UNITS_ADD(self, axis, len(CONST.AXES)); return True
    def Add6D(self, vector):
        if not self._can_add(): return False
        self.mind.access.UNITS_ADD(self, self._axis(vector._values()), 6); return True
    def _apply_adjustment(self, values):
        # The C# switch returns after its first matching axis; preserve that
        # behaviour for parity instead of updating every axis at once.
        value = values[0]
        if value < CONST.MIN + CONST.MIN or value > CONST.MAX - CONST.MIN:
            self.mind.access.UNITS_REM(self)
        else:
            self.UIset("will", value)

    def Adjust2D(self):
        random_value = self.mind.rand.MyRandomDouble(10)[5]
        vector, func = self.ToVector2D(), GPTVector2D()
        direction = vector.Unit().ReverseUnit() if self.mind.down.Output(vector) < 0 else vector.Unit()
        updated = func.Add(vector, func.Mul(direction, random_value * CONST.ETA))
        if updated.xx <= CONST.MIN: updated.xx = CONST.MIN
        if updated.yy >= CONST.MAX: updated.yy = CONST.MAX
        self._apply_adjustment((updated.xx, updated.yy))

    def Adjust6D(self):
        random_value = self.mind.rand.MyRandomDouble(10)[5]
        vector2, vector6, func = self.ToVector2D(), self.ToVector6D(), GPTVector6D()
        direction = vector6.Unit().ReverseUnit() if self.mind.down.Output(vector2) < 0 else vector6.Unit()
        updated = func.Add(vector6, func.Mul(direction, random_value * CONST.ETA))
        updated.xx = max(CONST.MIN, updated.xx)
        for name in ("yy", "zz", "ww", "vv", "uu"):
            setattr(updated, name, min(CONST.MAX, getattr(updated, name)))
        self._apply_adjustment(updated._values())
    def IsUNIT(self): return self.unit_type == UNITTYPE.JUSTAUNIT
    def IsIDLE(self): return self.unit_type == UNITTYPE.IDLE
    def IsDECISION(self): return self.unit_type == UNITTYPE.LDECISION
    def IsQDECISION(self): return self.unit_type == UNITTYPE.QDECISION
    @property
    def Data(self):
        if self.data == "QUICK": return "QYES" if self.UIget("will") < 50.0 else "QNO"
        if self.data != "DATA": return self.data
        if self.mind is None: return ""
        subject = self.mind.hub.GetSubject(self)
        if subject == "init": return ""
        return Lookup().GetDATA(self.mind, Index(self.mind.mood.res_norm, self.mind), subject)
    @Data.setter
    def Data(self, value): self.data = value
    @property
    def IsValid(self):
        if self.mind is None: return True
        validation = self.mind.bot.validation
        if validation == VALIDATION.BOTH: return self.mind._internal.Valid(self) and self.mind._external.Valid(self)
        if validation == VALIDATION.EXTERNAL: return self.mind._external.Valid(self)
        if validation == VALIDATION.INTERNAL: return self.mind._internal.Valid(self)
        raise RuntimeError("IsValid")
    @property
    def Root(self):
        units = sorted(self.mind.hub.UnitsPerOccupasionc(), key=lambda unit: unit.created)
        if not units: return ""
        index = units.index(self) + 1 if self in units else 0
        return f"_{self.mind.hub.GetSubject(self)}{index}"
    @property
    def Variable(self):
        if self.mind is None: return 0.0
        mech_type = self.mind.mech.type
        if mech_type == MECHANICS.CIRCUIT_1_LOW: raise RuntimeError("UNIT, Variable")
        if mech_type in (MECHANICS.CIRCUIT_2_LOW, MECHANICS.BALLONHILL_LOW): result = LowZero(self.UIget("will"))
        elif mech_type == MECHANICS.TUGOFWAR_LOW: result = HighZero(self.UIget("will"))
        else: raise RuntimeError("UNIT, Variable")
        if CONST.transfer == TRANSFER.LOGISTIC: return self.mind.calc.Logistic(result * .1 - 5.0) * 100.0
        if CONST.transfer == TRANSFER.OTHER: raise NotImplementedError("UNIT, Variable 1")
        return result
